python/cudf/cudf/core/groupby/groupby.py

```diff
@@ -261,17 +262,17 @@ def __init__(
         """
         self.obj = obj
         self._as_index = as_index
-        self._by = by
+        self._by = by.copy(deep=True) if isinstance(by, _Grouping) else by
```

Comment:
Should we rather move the task of making a deep copy to the caller's of this constructor?

Reply:
Can you discuss a bit more what this would mean? How many places call this code? If you delegate this task to the callers of the constructor, are there places where the deep copy _isn't_ needed?

Reply:
I don't know how this code really works, but there is something fishy about just setting `_Grouping._obj` to the new `self.obj` since `_Grouping._key_columns` and `_Grouping.names` could have been derived from the old `self.obj`

Reply:
@bdice Made a change such that only the caller's take the responsibility to make deep copies: https://github.com/rapidsai/cudf/pull/13932/commits/e256d9660f1cf7d98e62565a9b04149d3664c023

Reply:
Sorry, I don't like this change in e256d96. `groupby(by=existing_grouper)` is now still unsafe by default and requires the caller to remember to make the copy. Since, AFAICT, they _always_ have to make the copy, let's do it for them. Additionally, they only have to make the copy when `by` is an existing `_Grouping` object (if it's just the keys, that's fine since those won't be modified), so to correctly determine when to copy needs inappropriate intimacy with the internal implementation of `GroupBy` (they need to inspect the `grouper._by` slot and check it's a `_Grouping` which feels like an internal implementation detail).

Reply:
> Since, AFAICT, they _always_ have to make the copy, let's do it for them.

Right, this is the crux of what I was asking earlier:

> If you delegate this task to the callers of the constructor, are there places where the deep copy _isn't_ needed?

I'm still parsing through this but my initial impression aligns with what @wence- is saying. Maybe revert e256d96, then? @wence- Are you generally happy with the state before e256d96?

Reply:
Looks like we are tending towards the initial version. I reverted the changes and moved copies to the constructor only when the `by` object is a `Grouping`.

Reply:
Thanks